Resolve MapConfig each time LocalRegionCache#cleanup is called

kennymacleod opened the pull request:
As discussed in #80, by re-resolving the `MapConfig` on every call to `LocalRegionCache#cleanup`, we will pick up any runtime changes to the config.

----

closes #80

kennymacleod commented on hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/local/LocalRegionCache.java:
```diff
@@ -87,12 +87,9 @@ public LocalRegionCache(final String name, final HazelcastInstance hazelcastInst
-        } catch (UnsupportedOperationException ignored) {
-            EmptyStatement.ignore(ignored);
-        }
+        this.config = lookupMapConfig(name, hazelcastInstance);
```
I would have liked to get rid of the `config` field entirely, but it's a `protected` field and so removing it might break something elsewhere. This class itself no longer uses it, though.

kennymacleod commented on hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/local/LocalRegionCache.java:
```diff
@@ -106,6 +103,14 @@ public LocalRegionCache(final String name, final HazelcastInstance hazelcastInst
 
+    private static MapConfig lookupMapConfig(String name, HazelcastInstance hazelcastInstance) {
+        try {
+            return hazelcastInstance != null ? hazelcastInstance.getConfig().getMapConfig(name) : null;
```
I've changed this from using `findMapConfig` to using `getMapConfig`, which has fewer weird side-effects but should still give us what we need.

devOpsHazelcast commented:
Can one of the admins verify this patch?

enozcan commented:
Hi @kennymacleod,

Thanks for the contribution!
Can you please apply these changes for `hazelcast-hibernate53` module and add test for the dynamic configuration change?

enozcan commented on hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/local/LocalRegionCache.java:
```diff
@@ -296,6 +301,7 @@ public Map asMap() {
     void cleanup() {
         final int maxSize;
         final long timeToLive;
+        final MapConfig config = lookupMapConfig(name, hazelcastInstance);
```
I just noticed an issue here. 

A map configuration cannot be changed after a cluster is up and running. However, new configurations can be added ([see doc here](https://docs.hazelcast.org/docs/latest/manual/html-single/#dynamically-adding-data-structure-configuration-on-a-cluster)). That means, in order to change a LocalRegion map config, you have to add another map config and change the local region's name. But the region itself is registered by its name and once its name is changed, this region is going to be unreachable. WDYT?

kennymacleod commented:
@enozcan sorry for the slow response. As far as dynamic updates to the `MapConfig` are concerned, you're quite right, this is generally not possible. In our own application, we've achieved this using some unpleasant low-level hackery, so we can do it, but I agree that this isn't somethng that people can generally do, so there's little value in catering for it.

I had a rethink (again), and I've introduced an `EvictionConfig` interface which defines the TTL and max size parameters, and by default this interface is backed by the `MapConfig` as before, but now there's also an overloaded constructor where an alternative implementation can be passed in.

I've implemented this for both hibernate5 and hibernate53 modules.

As for how to make use of this, an application that wanted to provide a custom `EvictionConfig` implementation would need to use a custom `RegionFactory` instead of the ones you provide, but those classes are much simpler than `LocalRegionCache` and easier to replace.

kennymacleod commented on hazelcast-hibernate53/src/main/java/com/hazelcast/hibernate/local/LocalRegionCache.java:
```diff
@@ -367,4 +384,42 @@ public int hashCode() {
+        /**
+         * @return the duration for which an item should live in the cache
+         */
+        Duration getTimeToLive();
```
The hib53 module is Java8, so can use the Duration type to make this slightly nicer

devOpsHazelcast commented:
[![CLA assistant check](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/pull/badge/not_signed)](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast-hibernate5?pullRequest=82) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast-hibernate5?pullRequest=82) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/check/hazelcast/hazelcast-hibernate5?pullRequest=82) it.</sub>

pivovarit commented:
Thanks @kennymacleod 

Looks good for me. Could you merge your fork with the master? 

There were some quite drastic changes made recently. I don't expect them to break anything, but better be safe than sorry.

kennymacleod commented:
@pivovarit would it make sense to merge to the 1.3.z branch (so that that a new 1.3.z can be released that can be used with hz 3), and then merge that forward to master?

pivovarit commented:
Sure! Actually, I did not notice that the PR is open against `1.3.z` :)

kennymacleod commented:
OK, cool. I don't have permission to merge, though, could you do the honours? It'll need omeone to re-run the release process also, to get us a 1.3.3 release.

pivovarit commented:
No worries, we'll take care of that

pivovarit commented:
thanks @kennymacleod 
I will backport(forwardport?) your changes and make you a co-author. Thanks again!

Mak-Sym commented:
@pivovarit thank you for merging these changes.
Do you know when we can expect new 1.3.3 version to be released?

pivovarit commented:
We're working on a few more things but I believe 1-2 weeks is realistic

Mak-Sym commented:
thank you!
